packages/datetime2/src/components/date-range-input3/dateRangeInput3.tsx

```diff
@@ -0,0 +1,942 @@
+            setRef(prevProps.endInputProps?.inputRef, null);
+            this.handleEndInputRef = refHandler(this, "endInputElement", this.props.endInputProps?.inputRef);
+            setRef(this.props.endInputProps?.inputRef, this.endInputElement);
+        }
+
+        const shouldFocusStartInput = this.shouldFocusInputRef(isStartInputFocused, this.startInputElement);
+        const shouldFocusEndInput = this.shouldFocusInputRef(isEndInputFocused, this.endInputElement);
+
+        if (shouldFocusStartInput) {
+            this.startInputElement?.focus();
+        } else if (shouldFocusEndInput) {
+            this.endInputElement?.focus();
+        }
+
+        if (isStartInputFocused && shouldSelectAfterUpdate) {
+            this.startInputElement?.select();
+        } else if (isEndInputFocused && shouldSelectAfterUpdate) {
+            this.endInputElement?.select();
+        }
+
+        let nextState: Partial<DateRangeInput3State> = {};
+
+        if (this.props.value !== prevProps.value) {
+            const [selectedStart, selectedEnd] = this.getInitialRange(this.props);
+            nextState = {
+                ...nextState,
+                selectedEnd,
+                selectedStart,
+            };
+        }
+
+        // cache the formatted date strings to avoid computing on each render.
+        if (this.props.minDate !== prevProps.minDate) {
+            const formattedMinDateString = this.getFormattedMinMaxDateString(this.props, "minDate");
+            nextState = { ...nextState, formattedMinDateString };
+        }
+        if (this.props.maxDate !== prevProps.maxDate) {
+            const formattedMaxDateString = this.getFormattedMinMaxDateString(this.props, "maxDate");
+            nextState = { ...nextState, formattedMaxDateString };
+        }
+
+        this.setState(nextState as DateRangeInput3State);
+    }
+
+    public render() {
+        const { selectedShortcutIndex } = this.state;
+        const { popoverProps = {}, popoverRef } = this.props;
+
+        const popoverContent = (
+            <DateRangePicker3
+                {...this.props}
+                selectedShortcutIndex={selectedShortcutIndex}
+                boundaryToModify={this.state.boundaryToModify}
+                onChange={this.handleDateRangePickerChange}
+                onShortcutChange={this.handleShortcutChange}
+                onHoverChange={this.handleDateRangePickerHoverChange}
+                value={this.getSelectedRange()}
+            />
+        );
+
+        // allow custom props for the popover and each input group, but pass them in an order that
+        // guarantees only some props are overridable.
+        return (
+            <Popover
+                isOpen={this.state.isOpen}
+                placement="bottom-start"
+                {...popoverProps}
+                autoFocus={false}
+                className={classNames(Classes.DATE_RANGE_INPUT, popoverProps.className, this.props.className)}
```

Comment:
nit: consider using `{ Classes as DatetimeClasses } from "@blueprintjs/datetime"` import here, to make it more clear that this is a datetime "v1" class name

```suggestion
                className={classNames(DatetimeClasses.DATE_RANGE_INPUT, popoverProps.className, this.props.className)}
```